docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+
+**Constraint Generator**: Generate a `TypeAliasExpansionConstraint` when visiting the `AstStatTypeFunction` node.
+
+**Constraint Solver**: Generate a new `ReduceConstraint` and append it to the list of unsolved constraints when reducing `TypeAliasExpansionConstraint`.
+
+**Reduction Constraint**: Serialize the function arguments into `lType` and execute the body of the function. Reduce to the deserialized version of the function return value.
+
+**lType**: Using the Lua API, create a library that interfaces between C++ and Luau and supports all of the function calls, serialization, and deserialization.
+
+## Drawback
+
+Type functions are handled at the analysis time, while `lType` is an implementation in the runtime. As a result, the proposed design causes the analysis time to be dependent on Luau's runtime to reduce user-defined type functions. This is generally discouraged as it is best to isolate the compile time, analysis time, and runtime from each other for the purpose of maintaining a clean separation of concerns, which helps minimize side effects and dependencies across different phases of the program execution and improves the modularity of the compiler. Overlaps between the analysis time and runtime can lead to code that is more complex and harder to manage, potentially increasing the risk of bugs and making the outcomes less predictable.
+
+The build / analysis times will also be negatively impacted by this feature as reducing type functions takes variable amount of time based on the program. The larger the type function, the longer it will take to reduce it in the constraint solver.
+
+## Alternatives
+
+### `table` Runtime Representation
+
+Currently, the runtime representation of types is a userdata called `lType`. Another representation is to use the already-existing type in Luau `table`; instead of serializing types into `lType`, we can serialize them into tables with predefined properties. For instance, the representation for a string singleton `"abc"` could be `{type = "stringSingleton", value = "abc"}`. So instead of writing:
+```luau
+type function isSingleton(t)
+    if t:isStringSingleton() then
+        return t
+    end
+end
+```
+developers could write:
+```luau
+type function isSingleton(t)
+    if t.type == "stringSingleton" then
+        return t
+    end
+end
+```
+
+One of the major advantages of using a table-based representation for types in Luau is that it eliminates the dependency on the virtual machine for type analysis. However, this design requires developers to have a deeper understanding of the runtime representation. For example, under the proposed design, a new table type can be created using `lType.new("table")`, while under the alternative design, tables must be declared with attributes: `{type = "table", props = {}, indexer = {}}`. This adds complexity to the developer experience and increases the chance of making syntax errors in their program.
```

Comment:
What prevents defining `lType.new` that returns a table?

Reply:
Nothing at all. If we did choose a table representation for types instead of a userdata, we would still provide a whole library to be linked into the VM created in analysis to support writing type functions effectively.

Reply:
As per this [comment](https://github.com/luau-lang/rfcs/pull/45#issuecomment-2187526252), `lType` userdata is only registered when reducing UDTFs, so any other uses of `lType` outside of type functions will be a call to invalid / nonexistant class.

Reply:
I'm just saying that this is kind of presented as a reason to go with userdata instead of tables.
There might be other reasons, but tables feel like a more natural primitive.
Userdata is a more rigid structure and reflection is likely to increase complexity (I don't think we can connect to TypeVariant directly here).